File: functions/src/callable/premium.ts
/**
 * 利用者のプレミアム期限を管理画面から動かす（docs/PREMIUM-DESIGN.md D4 / 6-3b）
 *
 * **決済サービスは入れない**（1 節）。期限が延びる道は 2 つだけで、
 * クーポンの引き換え（callable/coupons.ts）と、ここからの操作である。
 */
import { FieldValue, Timestamp, getFirestore } from 'firebase-admin/firestore';
import { getAuth } from 'firebase-admin/auth';
import { onCall } from 'firebase-functions/v2/https';

import { REGION, paths } from '../config';
import { extendedPremiumUntil } from '../domain/premium';
import { fail } from '../errors';
import { requireMonths } from './coupons';
import { requireSiteAdmin, requireString } from './access';

/**
 * 期限を延ばす／縮める（D4）。
 *
 * **月数は負でもよい。** 誤って配ったぶんを戻す手段が無いと、
 * クーポンを止めても付いてしまった期限だけが残る。
 *
 * 縮めても**既存のデータは消えない**（D3）。新しいリストを作れなくなり、
 * 容量の上限が既定へ戻るだけ（domain/quota.ts の resolveUserQuota）。
 */
export const extendPremium = onCall({ region: REGION }, async (request) => {
  requireSiteAdmin(request);
  const uid = requireString(request.data, 'uid', { maxLength: 200 });
  const months = requireMonths(
    (request.data as Record<string, unknown>)?.months,
    { allowNegative: true }
  );

  // **Auth 側で実在を確かめる。** users ドキュメントは初回ログイン時に
  // クライアントが作るため、「まだ無い＝居ない」ではない（index.ts の注記）。
  const user = await getAuth().getUser(uid).catch(() => null);
  if (!user) {
    throw fail('not-found', 'userNotFound');
  }

  const db = getFirestore();
  // **期限は本人だけの場所に置く（config.ts の userPrivate）。**
  // `users/{uid}` は誰でも ID 指定で読めるので、そこに置くと
  // 「誰がいつまでプレミアムか」が他の利用者にも見える。
  const privateRef = db.doc(paths.userPrivate(uid));
  const nowMs = Date.now();

  const untilMs = await db.runTransaction(async (tx) => {
    const snapshot = await tx.get(privateRef);
    const current = snapshot.data()?.premium?.until;
    const until = extendedPremiumUntil({
      currentUntilMs: current instanceof Timestamp ? current.toMillis() : null,
      months,
      nowMs,
    });

    tx.set(
      privateRef,
      {
        premium: {
          until: Timestamp.fromMillis(until),
          updatedAt: FieldValue.serverTimestamp(),
        },
      },
      { merge: true }
    );

    return until;
  });

  return { premiumUntil: untilMs };
});

File: lib/domain/premium.dart
/// プレミアムの判定（docs/PREMIUM-DESIGN.md 3.1）
///
/// **状態は `users/{uid}.premium.until` だけで表す。** 「プレミアムかどうか」の
/// 真偽値を別に持つと、期限が切れたときに 2 つが食い違う。
///
/// **サーバーにも同じ判定がある**（`functions/src/domain/premium.ts`）。
/// 実際に作成を止めるのはサーバーで、ここにあるのは**画面の出し分け**のため
/// だけの写しである。判定の正本はサーバー側であり、ここを緩めても
/// リストは作れない。
library;

/// プレミアムの判定。
class PremiumPolicy {
  const PremiumPolicy._();

  /// [until] の時点までプレミアム。まだ来ていなければ有効。
  ///
  /// [until] が null（`premium` が無い）なら、プレミアムでない
  /// （設計 7 節「読み取り時に既定へ倒す」）。
  ///
  /// **ちょうど同時刻は「切れている」として扱う。** サーバーの
  /// `until > now` と同じ向きにそろえる。片方だけ `>=` にすると、
  /// 境界の 1 瞬だけ画面にボタンが出るのに押すと断られる。
  static bool isActive(DateTime? until, {DateTime? now}) {
    if (until == null) return false;
    return until.isAfter(now ?? DateTime.now());
  }

  /// 入力されたクーポンコードを、送る形にそろえる。
  ///
  /// 前後の空白と、貼り付けで紛れ込む改行・タブだけを落とす。
  /// メールや紙から写すと末尾に空白が付き、そのままでは
  /// 正しいコードなのに「見つかりません」と出る。
  ///
  /// **大文字小文字はここで変えない。** 照合はサーバーが `codeHash` で
  /// 行い、その手前で大文字に揃えている
  /// （`functions/src/domain/coupon.ts` の `normalizeCouponCode`）。
  /// **同じ規則を 2 か所に持たない**——片方だけ直したときに、
  /// 正しいコードが通らなくなる。ここは、送る前に必ず要る整形
  /// （見えない空白を落とす）だけを受け持つ。
  static String normalizeCouponCode(String raw) =>
      raw.replaceAll(RegExp(r'\s'), '');
}
